### PR TITLE
Update README sections for installing and linking

### PR DESCRIPTION
**Pull request by yemreinci**
This PR updates the sections about how to install and link against the library.

### CODE REVIEW

**Review by ihsandemir — CHANGES_REQUESTED** [review]

File: README.md [hunk 1]
```diff
@@ -143,19 +145,115 @@ See the following for more information on Hazelcast IMDG:
...
+Pass the argument `-DCMAKE_INSTALL_PREFIX=/path/to/install` the first time you run `cmake` to configure 
+the installation directory:
+```sh
+cmake .. -DCMAKE_INSTALL_PREFIX=/path/to/install
+```
+
+#### 1.1.5.2. CMake configuration
+You can provide additional configuration options using the `-DVARIABLE=VALUE` syntax on the command line.
+Here are all the options that are supported:
+* `WITH_OPENSSL` : Set to `ON` to build the library with SSL support.
```
[the remark this answers is not in the document]
Comment (yemreinci) [reply README.md]: It has the default value documented. It is on the next line. 👇

File: README.md [hunk 2]
```diff
@@ -143,19 +145,115 @@ See the following for more information on Hazelcast IMDG:
...
+### 1.1.1. Requirements
+1. Linux, macOS or Windows
+2. A compiler that supports C++11
+3. [CMake](https://cmake.org) 3.10 or above 
+4. [Boost](https://www.boost.org) 1.71 or above 
+5. [OpenSSL](https://www.openssl.org) (optional)
+
+### 1.1.2. Downloading
+Go to the [releases](https://github.com/hazelcast/hazelcast-cpp-client/releases) page to 
+download the source code for the latest Hazelcast C++ client.
```
[the remark this answers is not in the document]
Comment (yemreinci) [reply README.md:163]: Added the "git clone checkout" alternative.

File: README.md [hunk 3]
```diff
@@ -143,19 +145,115 @@ See the following for more information on Hazelcast IMDG:
...
+Choose the one which suits your system the best.
+
+Follow the instructions for your platform:
+* [Linux and maxOS](#linux-and-macos-users)
+* [Windows](#windows-users)
+
+### 1.1.3. Linux and MacOS users
+Here is how you download and extract version 4.0.0:
+```sh
+curl -Lo hazelcast-cpp-client-4.0.0.tar.gz https://github.com/hazelcast/hazelcast-cpp-client/archive/v4.0.0.tar.gz
```
[the remark this answers is not in the document]
Comment (yemreinci) [reply README.md:175]: I believe it shouldn't. There are others ways to download the library and I wouldn't want to it crowd the requirements part with too many things. Any user who doesn't have curl installed can install it if the command fails.

File: README.md [hunk 4]
```diff
@@ -143,19 +145,115 @@ See the following for more information on Hazelcast IMDG:
...
+You also set the installation location in the configuration command above. 
+Pass the `-DCMAKE_INSTALL_PREFIX=C:\path\to\install` option to specify your desired installation location.
+
+Build and install:
+```bat
+cmake --build . --config Release
+cmake --build . --target install --config Release
+```
+The above commands will build and install the library with the `Release` configuration. 
+The install command may require administrator privileges depending on your install prefix.
```
[the remark this answers is not in the document]
Comment (yemreinci) [reply README.md:233]: Added a sentence stating that the user should make sure to pass the same `--config` parameter to both commands.

File: README.md [hunk 5]
```diff
@@ -143,19 +145,115 @@ See the following for more information on Hazelcast IMDG:
...
+* `WITH_OPENSSL` : Set to `ON` to build the library with SSL support. 
+This will require [OpenSSL](https://www.openssl.org) to be installed on your system. The default is `OFF`.
+* `BUILD_STATIC_LIB` : Set to `ON` or `OFF` depending on whether you want the static library. The default is `OFF`.
+* `BUILD_SHARED_LIB` : Set to `ON` or `OFF` depending on whether you want the shared library. The default is `ON`.
+* `DISABLE_LOGGING` : Setting this option to `ON` disables logging. The default is `OFF`.
+
+#### 1.1.5.3. Examples
+Build only the static library with SSL support:
+```sh
+cmake .. -DWITH_OPENSSL=ON -DBUILD_SHARED_LIB=OFF -DBUILD_STATIC_LIB=ON
```
[the remark this answers is not in the document]
Comment (yemreinci) [reply README.md:257]: Yes ?

[the remark this answers is not in the document]
Comment (yemreinci) [reply README.md:257]: This section isn't for building the examples.

File: README.md [hunk 6]
```diff
@@ -143,19 +145,115 @@ See the following for more information on Hazelcast IMDG:
...
+Then create and change into a new directory:
+```bat
+cd hazelcast-cpp-client-4.0.0
+mkdir build
+cd build
+```
+
+Run `cmake` to configure:
+```bat
+cmake -G "Visual Studio 16 2019" .. -DCMAKE_INSTALL_PREFIX=C:\HazelcastCxx
```
[the remark this answers is not in the document]
Comment (yemreinci) [reply README.md]: They are removed.

File: README.md [hunk 7]
```diff
@@ -143,19 +145,121 @@ See the following for more information on Hazelcast IMDG:
...
+download the source code for the latest Hazelcast C++ client.
+
+The releases page has both `tar.gz` and `zip` archives available. 
+Choose the one which suits your system the best.
+
+Follow the instructions for your platform:
+* [Linux and maxOS](#113-linux-and-macos-users)
+* [Windows](#114-windows-users)
+
+### 1.1.3. Linux and MacOS users
```
Comment (ihsandemir) [inline README.md:172]:
```suggestion
### 1.1.3. Linux and MacOS Users
```

File: README.md [hunk 8]
```diff
@@ -143,19 +145,121 @@ See the following for more information on Hazelcast IMDG:
...
-- Latest [Boost](https://www.boost.org/) Library (C++ client dependency)
-- Feature complete C++11 compiler (feature completeness for C++11 compiler and standard library are required. The minimum complete compiler for gcc is gcc 5.0)
+Run `cmake` again to build and install the library:
+```sh
+cmake --build .
+sudo cmake --build . --target install
+```
+See [this section](#1151-custom-install-location) for information on how to use a different installation location.
+
+### 1.1.4. Windows users
```
Comment (ihsandemir) [inline README.md:207]:
```suggestion
### 1.1.4. Windows Users
```

File: README.md [hunk 9]
```diff
@@ -143,19 +145,121 @@ See the following for more information on Hazelcast IMDG:
...
+cmake --build . --config Release
+cmake --build . --target install --config Release
+```
+The above commands will build and install the library with the `Release` configuration. 
+Make sure you pass the same `--config` option to both commands.
+
+The install command may require administrator privileges depending on your install prefix. 
+See [this section](#1151-custom-install-location) for information on how to use a different installation location.
+
+### 1.1.5. Advanced installation
```
Comment (ihsandemir) [inline README.md:236]:
```suggestion
### 1.1.5. Advanced Installation
```

File: README.md [hunk 10]
```diff
@@ -143,19 +145,121 @@ See the following for more information on Hazelcast IMDG:
...
+
+#### 1.1.5.1. Custom install location
```
Comment (ihsandemir) [inline README.md:238]:
```suggestion
#### 1.1.5.1. Custom Install Location
```

File: README.md [hunk 11]
```diff
@@ -143,19 +145,121 @@ See the following for more information on Hazelcast IMDG:
...
+Pass the argument `-DCMAKE_INSTALL_PREFIX=/path/to/install` the first time you run `cmake` to configure 
+the installation directory:
+```sh
+cmake .. -DCMAKE_INSTALL_PREFIX=/path/to/install
+```
+
+#### 1.1.5.2. CMake configuration
```
Comment (ihsandemir) [inline README.md:245]:
```suggestion
#### 1.1.5.2. CMake Configuration
```

File: README.md [hunk 12]
```diff
@@ -143,19 +145,121 @@ See the following for more information on Hazelcast IMDG:
...
+You can provide additional configuration options using the `-DVARIABLE=VALUE` syntax on the command line.
+Here are all the options that are supported:
+* `WITH_OPENSSL` : Set to `ON` to build the library with SSL support.
+This will require [OpenSSL](https://www.openssl.org) to be installed on your system. The default is `OFF`.
+* `BUILD_STATIC_LIB` : Set to `ON` or `OFF` depending on whether you want the static library. The default is `OFF`.
+* `BUILD_SHARED_LIB` : Set to `ON` or `OFF` depending on whether you want the shared library. The default is `ON`.
+* `DISABLE_LOGGING` : Setting this option to `ON` disables logging. The default is `OFF`.
+
+##### Example configuration commands
```
Comment (ihsandemir) [inline README.md:254]:
```suggestion
##### 1.1.5.2.1 Example Configuration Commands
```

File: README.md [hunk 13]
```diff
@@ -229,120 +333,71 @@ If you want to add a `Portable` class, you should use `<portable-factories>` ins
...
-Note: If you want to compile the library for your specific environment rather than using the release binary, see section [Reproducing Released Libraries From Source](#82-reproducing-released-libraries).
+If you are using CMake, see the section for [CMake users](#131-cmake-users).
 
-Download the latest Hazelcast C++ client library from [Hazelcast C++ Client Website](https://hazelcast.org/clients/cplusplus/). You need to download the zip file for your platform. For Linux and Windows, 32- and 64-bit libraries exist. For MacOS, there is only 64-bit version. 
+If you are not, then read the instructions specific to your platform:
+* [Linux and MacOS](#132-linux-and-macos-users)
+* [Windows](#133-windows-users)
 
-Unzip the file. Following is the directory structure for Linux 64-bit zip. The structure is similar for the other C++ client distributions.
+#### 1.3.1. CMake users
```
Comment (ihsandemir) [inline README.md:344]:
```suggestion
#### 1.3.1. CMake Users
```

File: README.md [hunk 14]
```diff
@@ -229,120 +333,71 @@ If you want to add a `Portable` class, you should use `<portable-factories>` ins
...
-	find_package(OpenSSL REQUIRED)
-	include_directories(${OPENSSL_INCLUDE_DIR})
-	link_libraries(${OPENSSL_SSL_LIBRARIES} ${OPENSSL_CRYPTO_LIBRARIES})
+The package name depends on the specific library type you want to use. 
+Options are `hazelcastcxx`, `hazelcastcxx_ssl`, `hazelcastcxx_static`, and `hazelcastcxx_ssl_static`.
+
+Make sure you add the installation prefix of the client library to `CMAKE_PREFIX_PATH` 
+if you are using a custom installation location. 
+
+#### 1.3.2. Linux and MacOS users
```
Comment (ihsandemir) [inline README.md:359]:
```suggestion
#### 1.3.2. Linux and MacOS Users
```

File: README.md [hunk 15]
```diff
@@ -229,120 +333,71 @@ If you want to add a `Portable` class, you should use `<portable-factories>` ins
...
-The test sources depend on the Boost development libraries (thread and chrono), OpenSSL (only needed if compiling with SSL support) and [apache thrift](https://github.com/apache/thrift) (needed for communications to remote controller).
-
-The easiest way to compile and run the tests is to use the test scripts. The scripts are :
-- `testLinuxSingleCase.sh` for linux and MAC OS.
-- `testWindowsSingleCase.bat` for Windows.
-
-The scripts accept the following options:
-
-`testLinuxSingleCase.sh <32|64> <SHARED|STATIC> <Debug|Release> [COMPILE_WITHOUT_SSL]`
+#### 1.3.3. Windows users
```
Comment (ihsandemir) [inline README.md:387]:
```suggestion
#### 1.3.3. Windows Users
```

File: README.md [hunk 16]
```diff
@@ -3,18 +3,20 @@
 
 * [Introduction](#introduction)
 * [1. Getting Started](#1-getting-started)
-  * [1.1. Requirements](#11-requirements)
+  * [1.1. Installing](#11-installing)
+    * [1.1.1. Requirements](#111-requirements)
+    * [1.1.2. Downloading](#112-downloading)
+    * [1.1.3. Linux and MacOS users](#113-linux-and-macos-users)
+    * [1.1.4. Windows users](#114-windows-users)
+    * [1.1.5. Advanced installation](#115-advanced-installation)
```
Comment (ihsandemir) [inline README.md:11]: Will you add subsections 1.1.5.1 and 1.1.5.2 to ToC?

**Review by ihsandemir — CHANGES_REQUESTED** [review]

**Review by ihsandemir — APPROVED** [review]

**Pull request merged**